gateway/src/metrics.rs

```
@@ -0,0 +1,1069 @@
+            tx.send(rack_id).expect("why has the sensor-poller task gone away?")
+        }
+        // ignoring duplicate attempt to set the rack ID...
+    }
+
+    pub async fn update_server_addrs(&self, new_addrs: &[SocketAddrV6]) {
+        self.addrs_tx.send_if_modified(|current_addrs| {
+            if current_addrs.len() == new_addrs.len()
+                // N.B. that we could make this "faster" with a `HashSet`,
+                // but...the size of this Vec of addresses is probably going to
+                // two or three items, max, so the linear scan actually probably
+                // outperforms it...
+                && current_addrs.iter().all(|addr| new_addrs.contains(addr))
+            {
+                return false;
+            }
+
+            // Reuse existing `Vec` capacity if possible.This is almost
+            // certainly not performance-critical, but it makes me feel happy.
+            current_addrs.clear();
+            current_addrs.extend_from_slice(new_addrs);
+            true
+        });
+    }
+}
+
+impl Drop for Metrics {
+    fn drop(&mut self) {
+        // Clean up our children on drop.
+        self.server.abort();
+        self.pollers.abort();
+    }
+}
+
+impl MetricsConfig {
+    fn oximeter_collection_interval(&self) -> Duration {
+        Duration::from_secs(self.oximeter_collection_interval_secs as u64)
+    }
+
+    /// Returns the number of sample chunks from individual SPs to buffer.
+    fn sample_channel_capacity(&self) -> usize {
+        // Roughly how many times will we poll SPs for each metrics collection
+        // interval?
+        let polls_per_metrics_interval = {
+            let collection_interval_ms: usize = self
+                .oximeter_collection_interval()
+                .as_millis()
+                .try_into()
+                .expect("your oximeter collection interval is way too big...");
+            collection_interval_ms / self.sp_poll_interval_ms
```

Do we need to guard against someone trying to run a config with the poll interval set to 0?

---

Yeah, probably. I wonder if we should just not start the SP polling task if it's set to 0...or we could just return an error.

---

I'd go with returning an error. Arguably even an `.expect()` would be fine, given the one just above this for a different kind of invalid timeout.